Arm/Memory/Separate.lean

```diff
@@ -115,6 +116,505 @@ example : mem_subset (BitVec.ofNat 64 (2^64 - 1)) 0#64 0#64 (BitVec.ofNat 64 (2^
+    have h₂ : a.toNat + n.toNat - 2^64 < 2^64 := by omega
+    have hmod : (a.toNat + n.toNat) % 18446744073709551616 = a.toNat + n.toNat - 2^64 := by
+      rw [Nat.mod_eq_sub_mod]
+      rw [Nat.mod_eq_of_lt h₂]
+      omega
+    rw [hmod] at h
+    omega
+
+/--
+If we express a memory region as `[a..(a+n)]` for `(n : Nat)`,
+and this memory region is legal, then we could not have had any wraparound.
+Thus, it must be the case that (a + n).toNat = a.toNat + n
+-/
+theorem toNat_add_distrib_of_mem_legal_of_lt
+    (h : mem_legal a (a + n)) :
+    (a + n).toNat = a.toNat + n.toNat := by
+  simp only [add_def, toNat_ofNat, Nat.reducePow]
+  rw [Nat.mod_eq_of_lt]
+  apply add_lt_of_mem_legal_of_lt h
+
 end Separate
 
+
+/-#
+
+### New Memory Model
+
+The new memory model is different from the old one in two ways:
+
+1. It uses (base pointer, length) to keep track of memory regions instead of closed inntervals of [pointer 1, pointer 2].
+2. To faciliatate the new representation, it bakes in the assumption that the memory region is legal
+   (i.e. no wraparound).
+3. More softly, it tries to keep reasoning in terms of `Nat` rather than `BitVec` in order to allow easier
+   automation via `omega` for proving disjointedness / subset assumptions.
+
+All of the new definitions are named after the old definitions with a prime (') after their name.
+For robustness (and confidence), we plan to prove theorems that establish the equivalence of the old and new memory models.
+-/
+section NewDefitions
+
+/--
+`mem_legal' a n` witnessses that `(a + n)` does not overflow, and thus `[a..a+n)` is a valid range
+of memory. Note that the interval is left closed, right open, and thus `n` is the number of bytes in the memory range.
+-/
+def mem_legal' (a : BitVec 64) (n : Nat) : Prop :=
+  a.toNat + n ≤ 2^64
+
+/--
+@bollu: have proof automation exploit this.
+Equation lemma for `mem_legal'`.
+-/
+theorem mem_legal'_def (h : mem_legal' a n) : a.toNat + n ≤ 2^64 := h
+
+/--
+The maximum size of the range we can choose to allocate is 2^64.
+@bollu: have proof automation exploit this.
+-/
+theorem mem_legal'.size_le_two_pow (h : mem_legal' a n) : n ≤ 2 ^ 64 := by
+  rw [mem_legal'] at h
+  omega
+
+/--
+Legal in the new sense implies legal in the old sense.
+-/
+theorem mem_legal_of_mem_legal' (h : mem_legal' a n) :
+    mem_legal a (a + (BitVec.ofNat 64 (n - 1))) := by
+  simp only [mem_legal', mem_legal, BitVec.le_def] at h ⊢
+  rw [BitVec.toNat_add_eq_toNat_add_toNat]
+  simp only [BitVec.toNat_ofNat, Nat.reducePow, Nat.le_add_right, decide_True]
+  rw [BitVec.toNat_ofNat]
+  rw [Nat.mod_eq_of_lt (by omega)]
+  omega
+
+
+/--
+Legal in the new sense implies legal in the old sense.
+Note that the subtraction could also have been written as `(b - a).toNat + 1`
+-/
+theorem mem_legal'_of_mem_legal (h: mem_legal a b) : mem_legal' a (b.toNat - a.toNat + 1) := by
+  simp only [mem_legal, decide_eq_true_eq] at h
+  rw [mem_legal']
+  rw [BitVec.le_def] at h
+  omega
```

Comment:
These feel like they should be combined into a `mem_legal_iff_mem_legal'` (or prime iff non-prime) lemma

Reply:
It's not so simple, because the bounds of `mem_legal` look slightly different, so it's not clear what value such a theorem would have.

Reply:
But I proved it anyway!